File: javascript/704.binary-search/704.binary-search.js
/*
 * @lc app=leetcode id=704 lang=javascript
 *
 * [704] Binary Search
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function(nums, target) {
    let left = 0;
    let right = nums.length - 1;
    let res = -1;

    while (right - left >= 0) {
        let m = Math.trunc((right + left) / 2);

        if (nums[m] === target) {
            res = m;
            break;
        }

        if (right === left) break;

        if (nums[m] < target) {
            left = m + 1;
        } else {
            right = m;
        }
    }

    return res
};
// @lc code=end

File: python/704.binary-search/704.binary-search.py
#
# @lc app=leetcode id=704 lang=python3
#
# [704] Binary Search
#

# @lc code=start
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        if len(nums) == 1:
            return 0 if nums[0] == target else -1

        l = 0
        r = len(nums) - 1

        while l <= r:
            p = l + ((r - l) // 2)
            center = nums[p]
            if center == target:
                return p

            elif center < target:
                l = p + 1
                continue
            elif center > target:
                r = p - 1

        return -1
# @lc code=end
